base/linalg/dense.jl

```diff
@@ -309,17 +309,60 @@ kron(a::AbstractVector, b::AbstractVector)=vec(kron(reshape(a,length(a),1),resha
 kron(a::AbstractMatrix, b::AbstractVector)=kron(a,reshape(b,length(b),1))
 kron(a::AbstractVector, b::AbstractMatrix)=kron(reshape(a,length(a),1),b)
 
-^(A::AbstractMatrix, p::Integer) = p < 0 ? inv(A^-p) : Base.power_by_squaring(A,p)
+# Matrix power
+function ^{T}(A::AbstractMatrix{T}, p::Real)
+    # non-inversible matrix to negative power is undefined.
+    if (p < 0) && (det(A) == 0)
+        throw(ArgumentError("Negative powers of singular matrix are not defined"))
+    end
 
-function ^(A::AbstractMatrix, p::Number)
+    # For integer powers, use repeated squaring
     if isinteger(p)
-        return A^Integer(real(p))
+        return p < 0 ? Base.power_by_squaring(inv(A),-Integer(p)) : Base.power_by_squaring(A,Integer(p))
```

Comment:
if `A` is integer, what should `A^2` produce?
On your second comment, you think we should always use diagonalization, even `p` is integer?